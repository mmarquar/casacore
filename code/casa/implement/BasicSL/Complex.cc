//# complex.cc:
//# Copyright (C) 1999
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

/* 
Copyright (C) 1988 Free Software Foundation
    written by Doug Lea (dl@rocky.oswego.edu)

This file is part of the GNU C++ Library.  This library is free
software; you can redistribute it and/or modify it under the terms of
the GNU Library General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your
option) any later version.  This library is distributed in the hope
that it will be useful, but WITHOUT ANY WARRANTY; without even the
implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.  See the GNU Library General Public License for more details.
You should have received a copy of the GNU Library General Public
License along with this library; if not, write to the Free Software
Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
*/

#ifdef __GNUG__
#pragma implementation
#endif
#include <aips/Mathematics/Complex.h>
#include <aips/Mathematics/Constants.h>

//
// FROM builtin.h in libg++
//
typedef void (*one_arg_error_handler_t)(const char*);


// error handling

void default_Complex_error_handler(const char* msg)
{
  cerr << "Fatal Complex arithmetic error. " << msg << "\n";
  exit(1);
}

one_arg_error_handler_t Complex_error_handler = default_Complex_error_handler;

one_arg_error_handler_t set_Complex_error_handler(one_arg_error_handler_t f)
{
  one_arg_error_handler_t old = Complex_error_handler;
  Complex_error_handler = f;
  return old;
}

Bool near(G_COMPLEX(float) val1, G_COMPLEX(float) val2, double tol) {
  if (tol <= 0)
    return ToBool(val1 == val2);
  if (val1 == val2) return True;
  if (val1 == 0)
    return ToBool(abs(val2) <= (1+tol)*FLT_MIN);
  else if (val2 == 0)
    return ToBool(abs(val1) <= (1+tol)*FLT_MIN);
  return ToBool(abs(val1-val2) <= tol*max(abs(val1),abs(val2)));
}

Bool near(G_COMPLEX(double) val1, G_COMPLEX(double) val2, double tol) {
  if (tol <= 0)
    return ToBool(val1 == val2);
  if (val1 == val2) return True;
  if (val1 == 0)
    return ToBool(abs(val2) <= (1+tol)*DBL_MIN);
  else if (val2 == 0)
    return ToBool(abs(val1) <= (1+tol)*DBL_MIN);
  return ToBool(abs(val1-val2) <= tol*max(abs(val1),abs(val2)));
}

Bool nearAbs(G_COMPLEX(float) val1, G_COMPLEX(float) val2, double tol) {
  return ToBool(abs(val2 - val1) <= tol);
}

Bool nearAbs(G_COMPLEX(double) val1, G_COMPLEX(double) val2, double tol) {
  return ToBool(abs(val2 - val1) <= tol);
}

g_implement2(G_COMPLEX,double,G_COMPLEX_CTOR_OP_IMP(double,int) G_COMPLEX_CTOR_OP_IMP(double,float) G_COMPLEX_ASSIGN_OP_IMP(double,float) G_COMPLEX_ASSIGN_OP_IMP(double,int) G_COMPLEX_OPEQ_IMP(double,float) G_COMPLEX_OPEQ_IMP(double,int))
g_implement2(G_COMPLEX,float,G_COMPLEX_CTOR_OP_IMP(float,int) G_COMPLEX_CTOR_OP_IMP(float,double) G_COMPLEX_ASSIGN_OP_IMP(float,double) G_COMPLEX_ASSIGN_OP_IMP(float,int) G_COMPLEX_OPEQ_IMP(float,int) G_COMPLEX_OPEQ_IMP(float,double))
g_implement2(G_COMPLEX,int,G_COMPLEX_CTOR_OP_IMP(int,float) G_COMPLEX_CTOR_OP_IMP(int,double) G_COMPLEX_ASSIGN_OP_IMP(int,double) G_COMPLEX_ASSIGN_OP_IMP(int,float) G_COMPLEX_OPEQ_IMP(int,float) G_COMPLEX_OPEQ_IMP(int,double))

#if defined(COMPLEX_STUPID_COMPILER)
//#***************************************************************************
//#********************* Should not be needed, casts *************************
//#********************* should take care of these   *************************
//#***************************************************************************
G_COMPLEX_DO_BIN_OP_IMP
#endif

Bool isNaN(const Complex &val)
{
    return ToBool(isNaN(val.real()) || isNaN(val.imag()));
}

Bool isNaN(const DComplex &val)
{
    return ToBool(isNaN(val.real()) || isNaN(val.imag()));
}

void setNaN(Complex &val)
{
    setNaN(val.real()); setNaN(val.imag());
}

void setNaN(DComplex &val)
{
    setNaN(val.real()); setNaN(val.imag());
}
