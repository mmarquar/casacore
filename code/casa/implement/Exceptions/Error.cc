//# Error.cc: Base class for all AIPS++ errors (templated classes)
//# Copyright (C) 1993,1994,1995,1996
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Exceptions/Error.h>
#include <iostream.h>


//#  This template is for index errors  where the Index can be any type. It
//#  is derived from the class IndexError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(indexError);
rtti_imp_mbrf_a1(indexError);

template<class t>
indexError<t>::indexError(t oI, const Char *str) : oIndex(oI), IndexError(str)
{}
template<class t>
indexError<t>::indexError(t oI, const String &str) : oIndex(oI), IndexError(str)
{}

template<class t>
indexError<t>::indexError(ExcpError *excp) : IndexError(excp)
{
    indexError<t> *tmp;
    PCAST(tmp,indexError<t>,excp);
    if (tmp) {
        oIndex = (*tmp).oIndex;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
indexError<t>::~indexError()
    {}

//#  This template is for duplicate key errors  where the Key can be any type. It
//#  is derived from the class DuplError. This derivation allows the user
//#  to catch all index errors with one catch statement. The remainder of
//#  of the class is very similar to the AipsError class.
rtti_imp_init_a1(duplError);
rtti_imp_mbrf_a1(duplError);

template<class t>
duplError<t>::duplError(ExcpError *excp) : DuplError(excp)
{
    duplError<t> *tmp;
    PCAST(tmp,duplError<t>,excp);
    if (tmp) {
        oKey = (*tmp).oKey;
        _equal = True;
    } else {
        _equal = False;
    }
}

template<class t>
duplError<t>::~duplError()
    {}


template<class t>
duplError<t>::duplError(t oI, const Char *str) : oKey(oI), DuplError(str)
{}
template<class t>
duplError<t>::duplError(t oI, const String &str) : oKey(oI), DuplError(str)
{}

