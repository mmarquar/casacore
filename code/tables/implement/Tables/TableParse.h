//# TableParse.h: Classes to hold results from table grammar parser
//# Copyright (C) 1994,1995,1997
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#if !defined(AIPS_TABLEPARSE_H)
#define AIPS_TABLEPARSE_H

#if defined(_AIX)
#pragma implementation ("TableParse.cc")
#endif 

//# Includes
#include <aips/aips.h>
#include <aips/Tables/Table.h>
#include <aips/Utilities/String.h>
#include <aips/Utilities/Sort.h>
#include <aips/Containers/List.h>
#include <aips/Containers/Block.h>
#include <aips/Tables/ExprNode.h>

//# Forward Declarations
class TableExprNode;
class AipsIO;
template<class T> class Vector;


// <summary>
// Helper class for class TableParse
// </summary>

// <use visibility=local>

// <reviewed reviewer="" date="" tests="">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
//   <li> TableParse
// </prerequisite>

// <etymology>
// TableParseVal holds a value found when analyzing a table command.
// </etymology>

// <synopsis> 
// A table command is lexically analyzed via flex.
// An object of this class is used to hold a value (like a name
// or a literal) for later use in the parser code.
// </synopsis> 


class TableParseVal
{
public:
    Int      type;              //# i=Int, f=double, c=DComplex, s=String
    String   str;               //# string literal; table name; field name
    Int      ival;              //# integer literal
    double   dval[2];           //# double/DComplex literal

    // Make a new TableParseVal object (for the flex actions).
    static TableParseVal* makeValue();
};




// <summary>
// Helper class for class TableParse
// </summary>

// <use visibility=local>

// <reviewed reviewer="" date="" tests="">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
//   <li> TableParse
// </prerequisite>

// <etymology>
// TableParseSort holds a sort expression and order.
// </etymology>

// <synopsis> 
// A table command is parsed.
// An object of this class is used to hold the sort expression
// and sort order.
// </synopsis> 


class TableParseSort
{
public:
    // Construct from a given expression and for the given order.
    TableParseSort (const TableExprNode&, Sort::Order);

    ~TableParseSort();

    // Get the expression node.
    const TableExprNode& node() const;

    // Get the sort order.
    Sort::Order order() const;

private:
    TableExprNode node_p;
    Sort::Order   order_p;
};




// <summary>
// Class to hold values from table grammar parser
// </summary>

// <use visibility=local>

// <reviewed reviewer="" date="" tests="">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
// </prerequisite>

// <etymology>
// TableParse is the class used to parse a table command.
// </etymology>

// <synopsis> 
// TableParse is used by the parser of table select statements.
// The parser is written in Bison and Flex in files TableGram.y and .l.
// The statements in there use the routines in this file to act
// upon a reduced rule.
// Since multiple tables can be given (with a shorthand), the table
// names are stored in a list. The variable names can be qualified
// by the table name and will be looked up in the appropriate table.
//
// A select command is similar to SQL and can look like:
//    SELECT columns FROM tab1 sh1, tab2 sh2, tab3 WHERE
//           sh1.field == 3*sh1.field2 ... ORDERBY columns GIVING table
// This is described in more detail in TableGram.l.
//
// The class TableParse only contains information about a table
// used in the table command. Global variables (like a list and a vector)
// are used in TableParse.cc to hold further information.
//
// Global functions are used to operate on the information.
// The main function is the global function tableCommand.
// This function executes the given command and returns the resulting table.
// This is, in fact, the only function to be used by a user.
// </synopsis> 

// <motivation>
// It is necessary to be able to give a table select command in ASCII.
// This can be used in a CLI or in the table browser to get a subset
// of a table or to sort a table.
// </motivation>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>


class TableParse
{
// Dummy AipsIO routines; they are needed for the List container.
// <group>
friend AipsIO& operator<< (AipsIO&, const TableParse&);
friend AipsIO& operator>> (AipsIO&, TableParse&);
// </group>

// Parse and execute the given command.
// It will open (and close) all tables needed.
// It returns the resulting table. 
friend Table tableCommand (const String& command);

// Parse and execute the command.
// It will open (and close) all tables needed.
// It returns the resulting table and the selected column names.
friend Table tableCommand (const String& command,
			   Vector<String>& columnNames);

public:
    // Default constructor for List container class.
    TableParse ();

    // Copy constructor (copy semantics).
    TableParse (const TableParse&);

    // Assignment (copy semantics).
    TableParse& operator= (const TableParse&);

    // Associate the table and the shorthand.
    TableParse (const Table& table, const String& shorthand);

    // Test if shorthand matches.
    Bool test (const String& shortHand) const;

    // Get the shorthand.
    const String& shorthand() const;

    // Get table object.
    const Table& table() const;

private:
    String  shorthand_p;
    Table   table_p;
};


// <summary>
// Select-class for flex/bison scanner/parser for TableParse
// </summary>

// <use visibility=local>

// <reviewed reviewer="" date="" tests="">
// </reviewed>

// <prerequisite>
//# Classes you should understand before using this one.
//  <li> TableParse
//  <li> TableGram.l and .y  (flex and bison grammar)
// </prerequisite>

// <synopsis> 
// This class is needed for the the actions in the flex scanner
// and bison parser.
// This stores the information by constructing TableParse objects
// as needed and storing them in a List.
// </synopsis> 

// <motivation>
// It is necessary to be able to give a table select command in ASCII.
// This can be used in a CLI or in the table browser to get a subset
// of a table or to sort a table.
// </motivation>

//# <todo asof="$DATE:$">
//# A List of bugs, limitations, extensions or planned refinements.
//# </todo>


class TableParseSelect
{
public:

    // Default constructor.
    TableParseSelect();

    // Destructor.
    ~TableParseSelect();

    // Execute the selection.
    void doSelect (const TableExprNode*);

    // Sort the resulting table.
    // After the sort it deletes the list and the elements in it.
    void doSort (PtrBlock<TableParseSort*>* sortList);

    // Add a table name to the linked list.
    void addTable (const String& name, const String& shorthand);

    // Find the field name and create a TableExprNode from it.
    //# String has to be non-const due to use of String::before/after.
    TableExprNode handleName (String& str, Bool isArray);

    // Handle an array.
    TableExprNode handleArray (String& str, Block<TableExprNode>& indices);

    // Handle a function.
    TableExprNode handleFunc (const String& name,
			      Block<TableExprNode>& arguments);

    // Create a TableExprNode from a literal.
    TableExprNode handleLiteral (TableParseVal*);

    // Add a column to the list of selected column names.
    //# String has to be non-const due to use of String::before/after.
    void handleColumn (String& name);

    // Handle the name given in a GIVING clause.
    void handleGiving (const String&);

    // functions to 'get' stored data
    // <group>
//    ListIter<TableParse>& getList();
//    Block<String>*        getSkey();
//    Block<Int>*           getSord();
    const Block<String>&  getColumnNames() const;
    const Table&          getTable() const;
    const String&         getTableName() const;
    Bool                  getNothingDone() const;
    // </group>

    // Create a new TableParseSelect-object and put it on the stack (block)
    // If currentTableParseSelect-object does not exist, create it.
    static void newSelect();

    // Delete currentTableParseSelect object and get the next one.
    // The last one is not deleted; this is done by the tableCommand function.
    // In that way tableCommand can extract information from it
    // (e.g. the selected columns).
    static void deleteSelect();

    // Get current TableParseSelect object.
    static TableParseSelect*& currentSelect();


private:
    // Block of TableParseSelect objects (acts like a stack).
    static PtrBlock<TableParseSelect*> blockSelect_p;

    // Pointer to current TableParseSelect object (in Block above).
    // (Acts like pointer to top of stack).
    static TableParseSelect* currentSelect_p;

    //# Pointer to a linked list of TableParse objects.
    //# This is needed for the functions above, otherwise they have no
    //# way to communicate.
    List<TableParse>*     parseList_p;
    ListIter<TableParse>* parseIter_p;

    //# Block of selected column names.
    Block<String> columnNames_p;

    //# Name of the resulting table (from GIVING part).
    String resultName_p;

    //# Determine if selection or sorting is done.
    Bool nothingDone_p;
};



//# Implement the inline functions.
inline Bool TableParse::test (const String& str) const
    { return (shorthand_p == str  ?  True : False); }

inline const String& TableParse::shorthand() const
    { return shorthand_p; }

inline const Table& TableParse::table() const
    { return table_p; }


inline const Block<String>& TableParseSelect::getColumnNames() const
    { return columnNames_p; }

inline const String& TableParseSelect::getTableName() const
    { return resultName_p; }
 
inline Bool TableParseSelect::getNothingDone() const
    { return nothingDone_p; }

inline TableParseSelect*& TableParseSelect::currentSelect()
    { return TableParseSelect::currentSelect_p; }


inline const TableExprNode& TableParseSort::node() const
    { return node_p; }

inline Sort::Order TableParseSort::order() const
    { return order_p; }



#endif
