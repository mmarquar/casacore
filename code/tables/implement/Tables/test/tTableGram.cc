//# tTableGram.cc: This program tests table commands using TableGram/Parse
//# Copyright (C) 1994,1995,1996,1998,2000,2002,2003
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <aips/Tables/TableParse.h>
#include <aips/Tables/Table.h>
#include <aips/Tables/TableDesc.h>
#include <aips/Tables/TableColumn.h>
#include <aips/Arrays/Matrix.h>
#include <aips/Arrays/Vector.h>
#include <aips/BasicSL/Complex.h>
#include <aips/Exceptions/Error.h>
#include <aips/iostream.h>

// <summary>
// Test program for table commands from user interface
// </summary>

// This program tests table commands with an SQL-like grammar.
// The grammar is scanned and parsed using the flex/bison file TableGram.l/y
// and with the help of the class TableParse.
// It ask for commands until a "q" is given.
// When columns are selected, it will show their contents.


void seltab (const String&);
void docomm ();

int main (int argc, char** argv)
{
    try {
	if (argc > 1) {
	    seltab(argv[1]);
	}else{
	    docomm();
	}
    } catch (AipsError x) {
	cout << "\nCaught an exception: " << x.getMesg() << endl;
        return 1;
    } 
    return 0;               // successfully executed
}


// Ask and execute command till empty string is given.
void docomm()
{
    char comm[1025];
    while (True) {
	cout << "Table command (q=quit): ";
	cin.getline (comm, 1024);
	String str(comm);
	if (str == "q") 
	    break;
	try {
	    seltab (str);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;
	} 
    }
}


// Show the required columns.
// First test if they exist and contain scalars or arrays.
void showtab (const Table& tab, const Vector<String>& colnam)
{
    uInt nrcol = 0;
    PtrBlock<ROTableColumn*> tableColumns(colnam.nelements());
    uInt i;
    for (i=0; i<colnam.nelements(); i++) {
	if (! tab.tableDesc().isColumn (colnam(i))) {
	    cout << "Column " << colnam(i) << " does not exist" << endl;
	}else{
	    tableColumns[nrcol] = new ROTableColumn (tab, colnam(i));
	    if (! tableColumns[nrcol]->columnDesc().isScalar()
	    &&  ! tableColumns[nrcol]->columnDesc().isArray()) {
		cout << "Column " << colnam(i)
		     << " contains scalars nor arrays"
		     << endl;
		delete tableColumns[nrcol];
	    }else{
		nrcol++;
	    }
	}
    }
    if (nrcol == 0) {
	return;
    }

    for (i=0; i<tab.nrow(); i++) {
	for (uInt j=0; j<nrcol; j++) {
	    if (tableColumns[j]->columnDesc().isArray()) {
		cout << " " << tableColumns[j]->shape (i);
	    }else{
		switch (tableColumns[j]->columnDesc().dataType()) {
		case TpBool:
		    cout << " " << tableColumns[j]->asBool (i);
		    break;
		case TpString:
		    cout << " " << tableColumns[j]->asString (i);
		    break;
		case TpComplex:
		case TpDComplex:
		    cout << " " << tableColumns[j]->asDComplex (i);
		    break;
		default:
		    cout << " " << tableColumns[j]->asdouble (i);
		}
	    }
	}
	cout << endl;
    }

    for (i=0; i<nrcol; i++) {
	delete tableColumns[i];
    }
}


// Sort and select data.
void seltab (const String& str)
{
    uInt i;
    Table tab;
    Vector<String> vecstr;
    String cmd;
    cout << str << endl;
    // Select rows from the table.
    tab = tableCommand (str, vecstr, cmd);
    cout << "    has been executed" << endl;
    cout << "    " << cmd << " of " << tab.nrow() << " rows" << endl;
    // Show the selected column names.
    cout << vecstr.nelements() << " selected columns: ";
    for (i=0; i<vecstr.nelements(); i++) {
	cout << " " << vecstr(i);
    }
    cout << endl;

    // Show the contents of the columns.
    if (vecstr.nelements() > 0) {
	showtab (tab, vecstr);
    }
}
