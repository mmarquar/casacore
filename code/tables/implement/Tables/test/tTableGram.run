#!/bin/sh
#-----------------------------------------------------------------------------
# Script to test the TableGram and TableParse class.  All files generated will
# be deleted on exit.
#=============================================================================

# Use table tTable_2.data_v0 as the input by creating a link to its directory.
  rm -rf tTableGram_tmp.tab
  mkdir tTableGram_tmp.tab
  AIPSCODE=`echo $AIPSPATH | awk '{printf("%s/code",$1)}'`
  cp -r $AIPSCODE/aips/implement/Tables/test/tTable_2.data_v0/table.* tTableGram_tmp.tab
  chmod 644 tTableGram_tmp.tab/*

# Whitespace around * had to be removed to avoid file name expansion by shell.


tTableGram 'select ab,ac,ad,ae,af,ag into tTableGram_tmp.data2 from tTableGram_tmp.tab sh where all(ab>2) && (ae<10 || ae>11.0) && ag!= 10 + 1i orderby ac desc,ab'

tTableGram 'select distinct ab,ac,ad,ae,af,ag from tTableGram_tmp.data2'

tTableGram 'select all ab,ac,ad,ae,af,ag from tTableGram_tmp.data2 orderby af'

tTableGram 'select ab from tTableGram_tmp.tab where ab==2^1^2 || ab==-2^-1*8/-2*3'

tTableGram 'select ab,ac,af from tTableGram_tmp.tab where lower(af) == regex("v[01279]")'

tTableGram 'select ab,ac,af from tTableGram_tmp.tab where af == pattern("?{3,5,8}")'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab%1.5==0'

tTableGram 'select ab,ac from tTableGram_tmp.tab where arr1[1,1,1]>=10 && arr2[1,1,1]<120'

tTableGram 'select * from tTableGram_tmp.tab where arr1[1,1,1]>=10 && arr2[1,1,1]<120'

tTableGram 'select ab,ac from tTableGram_tmp.tab where arr1[1,1,1+ab%1]>=192 orderby ad desc'

tTableGram 'select ab,ac from tTableGram_tmp.tab where cos(2d0m) <= sin(-2d)*sin(-ab/180*pi()) + cos(-2deg)*cos(-ab/180*pi())*cos(3d - ac/180*pi())'

tTableGram 'select ab,ac,ad,ae,af,ag from tTableGram_tmp.tab where ab+ac+ad+ae+real(ag) >= year(31-12-60) + year("31Dec60") + month(1990/5/12) + day(date(1990/1/30/12h14m33.3)) - 3910'

tTableGram 'select ab,ac,af from tTableGram_tmp.tab where ab>5 orderby af desc, ac'

tTableGram 'select ab,ac,af from tTableGram_tmp.tab orderby arr1[1,1,1]'

tTableGram 'select ab,ac from tTableGram_tmp.tab orderby round(2*sin(ab)),ac desc'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab < mean([3:6,ab])'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab IN [select ac from tTableGram_tmp.tab where ab>4]'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab IN [:=2,4=:<6,7<:]'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab IN [select from tTableGram_tmp.tab where ab>4 giving [ac=:=ac+0.5]]'

tTableGram 'select ab,ac from tTableGram_tmp.tab where any(isnan(arr1)) || isnan(ab)'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab IN arr1'

tTableGram 'select ab,ac from tTableGram_tmp.tab where any(arr1 > 100  &&  arr1<200)'

tTableGram 'select ab,ac from tTableGram_tmp.tab where count(shape(arr1))==3 && count(ab)==1 && ndim(ac)==0 && isdefined(arr2)'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ab in ab'

tTableGram 'select ab,ac from tTableGram_tmp.tab where any(arr1 in ab)'

tTableGram 'select ab,ac from tTableGram_tmp.tab where (ab=ab)=T'

tTableGram 'select ab,ac from tTableGram_tmp.tab where (ab=ab)=F'

tTableGram 'select ab,ac from tTableGram_tmp.tab where rownumber()==rowid()+1'

tTableGram 'select ab,ac from [select from tTableGram_tmp.tab where ab > 4] where ab < 6'

tTableGram 'select ab,ac from [select from tTableGram_tmp.tab where ab > 4] TEMPTAB, tTableGram_tmp.tab where any([ab,ac] in [select ac from TEMPTAB])'

tTableGram 'select ab,ac from tTableGram_tmp.tab where ac in [select from tTableGram_tmp.tab where ac in 4:6:2 giving [rowid()]]'

tTableGram 'select ab from tTableGram_tmp.tab where min(maxs(arr1,[1+arr1[1,1,1]%2,3])) == 19'

tTableGram 'select ab from tTableGram_tmp.tab where min(1+maxs(arr1-1,1,3)) == 19'

tTableGram 'select ab from tTableGram_tmp.tab where sum(fractiles(arr1,0.5,[2:3])) == 21+shape(arr1)[1]*count(arr1)'

tTableGram 'select ab from tTableGram_tmp.tab where sum(ntrues(arr1%5==0,[1])) < 5'

tTableGram 'select ab from tTableGram_tmp.tab where all(anys(fmod(sums(arr1,1),5)==0,[2:4]))'


tTableGram 'update tTableGram_tmp.tab set ab=sum(arr1)+ac*2, arr1=arr1+2 where ac>3'
tTableGram 'select ab from tTableGram_tmp.tab'
