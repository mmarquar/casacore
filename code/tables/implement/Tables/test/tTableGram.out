select ab,ac,ad,ae,af,ag from tTableGram_tmp.link/tTable_2.data_v0 sh where ab>2 && (ae<10 || ae>11.0) && ag!= 10 + 1i orderby ac desc,ab giving tTableGram_tmp.data2
    has been executed
    5 rows selected
6 selected columns:  ab ac ad ae af ag
 9 10 11 12 V9 (11, 0)
 6 7 8 9 V6 (8, 0)
 5 6 7 8 V5 (7, 0)
 4 5 6 7 V4 (6, 0)
 3 4 5 6 V3 (5, 0)
select ab,ac,ad,ae,af,ag from tTableGram_tmp.data2
    has been executed
    5 rows selected
6 selected columns:  ab ac ad ae af ag
 9 10 11 12 V9 (11, 0)
 6 7 8 9 V6 (8, 0)
 5 6 7 8 V5 (7, 0)
 4 5 6 7 V4 (6, 0)
 3 4 5 6 V3 (5, 0)
select ab,ac,ad,ae,af,ag from tTableGram_tmp.data2 orderby af
    has been executed
    5 rows selected
6 selected columns:  ab ac ad ae af ag
 3 4 5 6 V3 (5, 0)
 4 5 6 7 V4 (6, 0)
 5 6 7 8 V5 (7, 0)
 6 7 8 9 V6 (8, 0)
 9 10 11 12 V9 (11, 0)
select ab from tTableGram_tmp.link/tTable_2.data_v0 where ab==2^1^2 || ab==-2^-1*8/-2*3
    has been executed
    2 rows selected
1 selected columns:  ab
 2
 6
select ab,ac,af from tTableGram_tmp.link/tTable_2.data_v0 where lower(af) == regex("v[01279]")
    has been executed
    5 rows selected
3 selected columns:  ab ac af
 0 1 V0
 1 2 V1
 2 3 V2
 7 8 V7
 9 10 V9
select ab,ac,af from tTableGram_tmp.link/tTable_2.data_v0 where af == pattern("?{3,5,8}")
    has been executed
    3 rows selected
3 selected columns:  ab ac af
 3 4 V3
 5 6 V5
 8 9 V8
select ab,ac from tTableGram_tmp.link/tTable_2.data_v0 where ab%1.5==0
    has been executed
    4 rows selected
2 selected columns:  ab ac
 0 1
 3 4
 6 7
 9 10
select ab,ac from tTableGram_tmp.link/tTable_2.data_v0 where arr1[1,1,1]>=10 && arr2[1,1,1]<120
    has been executed
    4 rows selected
2 selected columns:  ab ac
 1 2
 2 3
 3 4
 4 5
select from tTableGram_tmp.link/tTable_2.data_v0 where arr1[1,1,1]>=10 && arr2[1,1,1]<120
    has been executed
    4 rows selected
0 selected columns: 
select ab,ac from tTableGram_tmp.link/tTable_2.data_v0 where arr1[1,1,1+ab%1]>=192 orderby ad desc
    has been executed
    2 rows selected
2 selected columns:  ab ac
 9 10
 8 9
select ab,ac from tTableGram_tmp.link/tTable_2.data_v0 where cos(2d0m) <= sin(-2d)*sin(-ab/180*pi()) + cos(-2deg)*cos(-ab/180*pi())*cos(3d - ac/180*pi())
    has been executed
    3 rows selected
2 selected columns:  ab ac
 1 2
 2 3
 3 4
select ab,ac,ad,ae,af,ag from tTableGram_tmp.link/tTable_2.data_v0 where ab+ac+ad+ae+real(ag) >= year(31-12-60) + year("31Dec60") + month(1990/5/12) + day(date(1990/1/30/12h14m33.3)) - 3910
    has been executed
    2 rows selected
6 selected columns:  ab ac ad ae af ag
 8 9 10 11 V8 (10, 0)
 9 10 11 12 V9 (11, 0)
select ab,ac,af from tTableGram_tmp.link/tTable_2.data_v0 where ab>5 orderby af desc, ac
    has been executed
    4 rows selected
3 selected columns:  ab ac af
 9 10 V9
 8 9 V8
 7 8 V7
 6 7 V6
select ab,ac,af from tTableGram_tmp.link/tTable_2.data_v0 orderby arr1[1,1,1]
    has been executed
    10 rows selected
3 selected columns:  ab ac af
 0 1 V0
 1 2 V1
 2 3 V2
 3 4 V3
 4 5 V4
 5 6 V5
 6 7 V6
 7 8 V7
 8 9 V8
 9 10 V9
select ab,ac from tTableGram_tmp.link/tTable_2.data_v0 orderby round(2*sin(ab)),ac desc
    has been executed
    10 rows selected
2 selected columns:  ab ac
 5 6
 4 5
 6 7
 3 4
 0 1
 9 10
 7 8
 8 9
 2 3
 1 2
